### PR TITLE
Feature / MDS-1678; MDS-1679; General Search

### PR DESCRIPTION
**Pull request by nayr974**
This PR implements search generally for Core across multiple entities.

The search is based on individual text searches across many single columns, using an ILIKE filter to find matches that contains the search term (%term%), and then ranks them using PostgreSQL' trigram module (pg_trgm) for similarity to the search string. That ranking is then modified using some simple scoring  multipliers in the API to score exact matches, strings that start with the search term, then aggregate scores for each result across its multiple fields.

### CODE REVIEW

File: frontend/src/components/common/LinkButton.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    style={props.style}
+  >
+    {props.children}
+  </a>
+);
+
+LinkButton.propTypes = propTypes;
+LinkButton.defaultProps = defaultProps;
+
+export default LinkButton;
```
Comment (mmumby) [inline frontend/src/components/common/LinkButton.js]: This is a great component addition!

File: python-backend/app/api/search/search/resources/search.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    @api.marshal_with(SEARCH_RESULT_RETURN_MODEL, 200)
+    def get(self):
+        search_results = []
+        app = current_app._get_current_object()
+
+        search_term = request.args.get('search_term', None, type=str)
+        search_types = request.args.get('search_types', None, type=str)
+        search_types = search_types.split(',') if search_types else search_targets.keys()
+
+        reg_exp = regex.compile(r'\'.*?\' | ".*?" | \S+ ', regex.VERBOSE)
```
Comment (Jsyro) [inline python-backend/app/api/search/search/resources/search.py]: comment or helpful name on what this regex is supposed to do.

File: python-backend/app/api/search/search/resources/search.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    if getattr(item, value_field).lower().startswith(search_term.lower()):
+        score_multiplier = score_multiplier * 3
+
+    if getattr(item, value_field).lower() == search_term.lower():
+        score_multiplier = score_multiplier * 10
+
+    search_results.append(
+        SearchResult(
+            getattr(item, 'score') * score_multiplier, type, {
```
Comment (Jsyro) [inline python-backend/app/api/search/search/resources/search.py]: i don't believe there is any reason to use getattr unless you have defined a default value (3rd parameter), i think these should be item.value_field.

File: python-backend/app/api/search/search/resources/search.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                        func.similarity(column, term).label('score'),
+                        *type_config['entities']).filter(column.ilike(f'%{term}%'))
+                    if type_config['has_deleted_ind']:
+                        similarity = similarity.filter_by(deleted_ind=False)
+                    similarity = similarity.order_by(desc(func.similarity(column, term))).all()
+                    [
+                        append_result(search_results, search_term, type, item,
+                                      type_config['id_field'], type_config['value_field'],
+                                      type_config['score_multiplier']) for item in similarity
+                    ]
```
Comment (Jsyro) [inline python-backend/app/api/search/search/resources/search.py]: weird usage of list comprehension syntax, just loop with `for item in similarity`?

File: python-backend/app/api/search/search/resources/simple_search.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                        func.similarity(column, term).label('score'),
+                        *type_config['entities']).filter(column.ilike(f'%{term}%'))
+                    if type_config['has_deleted_ind']:
+                        similarity = similarity.filter_by(deleted_ind=False)
+                    similarity = similarity.order_by(desc(func.similarity(column, term))).all()
+                    [
+                        append_result(search_results, search_term, type, item,
+                                      type_config['id_field'], type_config['value_field'],
+                                      type_config['score_multiplier']) for item in similarity
+                    ]
```
Comment (Jsyro) [inline python-backend/app/api/search/search/resources/simple_search.py]: same execute_search and append_result?

Comment (Jsyro) [reply python-backend/app/api/search/search/resources/simple_search.py]: possibly move simple search, search, and search options into the same file, to share these methods.

File: python-backend/tests/search/resource/test_search_resource.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        f'/search?search_term={str(party.first_name)}', headers=auth_headers['full_auth_header'])
+    get_data = json.loads(get_resp.data.decode())
+    parties = get_data['search_results']['party']
+    assert len(parties) == 1
+    assert party.first_name in parties[0]['result']['name']
+    assert uuid.UUID(parties[0]['result']['party_guid']) == party.party_guid
+    assert len([
+        key for key, value in get_data['search_results'].items()
+        if key is not 'party' and len(value) is 0
+    ]) == 4
```
Comment (Jsyro) [inline python-backend/tests/search/resource/test_search_resource.py]: why 4? should this be dynamic? maybe a comment?

**Comment by Jsyro** [comment]
Backend looks good, 3 small comments. Great work,

File: python-backend/app/api/search/search/resources/search.py (hunk 2 above)
Comment (nayr974) [reply python-backend/app/api/search/search/resources/search.py]: Added

File: python-backend/app/api/search/search/resources/search.py (hunk 4 above)
Comment (nayr974) [reply python-backend/app/api/search/search/resources/search.py]: Changed

File: python-backend/tests/search/resource/test_search_resource.py (hunk 6 above)
Comment (nayr974) [reply python-backend/tests/search/resource/test_search_resource.py]: @JustinMacaulay thoughts?

File: python-backend/app/api/search/search/resources/simple_search.py (hunk 5 above)
Comment (nayr974) [reply python-backend/app/api/search/search/resources/simple_search.py]: Moved these to the helper class.

File: frontend/src/components/common/LinkButton.js (hunk 1 above)
Comment (nayr974) [reply frontend/src/components/common/LinkButton.js]: I agree! It seemed justified to wrap up that block that was being used all over.

File: frontend/src/actionCreators/searchActionCreator.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      dispatch(searchActions.storeSearchResults(response.data));
+      dispatch(hideLoading("modal"));
+    })
+    .catch(() => dispatch(error(reducerTypes.GET_SEARCH_RESULTS)))
+    .finally(() => dispatch(hideLoading("modal")));
+};
+
+export const fetchSearchBarResults = (searchTerm) => (dispatch) => {
+  dispatch(request(reducerTypes.GET_SEARCH_BAR_RESULTS));
+  dispatch(showLoading("modal"));
```
Comment (mmumby) [inline frontend/src/actionCreators/searchActionCreator.js]: Does this loading bar make sense? (in any of these actionCreators) it is modal specific loading, and I don't think you are using models for the search.

File: frontend/src/components/search/ContactResultsTable.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        showHeader={false}
+        pagination={false}
+        columns={columns}
+        dataSource={props.searchResults}
+      />
+      {props.showAdvancedLookup && (
+        <Link
+          style={{ float: "right", fontSize: "1.25rem" }}
+          className="padding-large--left"
+          to={router.CONTACT_HOME_PAGE.dynamicRoute(parseQuery(props.query))}
```
Comment (mmumby) [inline frontend/src/components/search/ContactResultsTable.js]: `router.CONTACT_HOME_PAGE.dynamicRoute` expects `page` and `per_page` params, I don't see these being set anywhere? Does this behave as expected?

File: frontend/src/components/search/ContactResultsTable.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  query: PropTypes.string.isRequired,
+  partyRelationshipTypeHash: PropTypes.objectOf(PropTypes.strings).isRequired,
+  showAdvancedLookup: PropTypes.bool.isRequired,
+};
+
+const defaultProps = {};
+
+const parseQuery = (query) => {
+  const emailExp = /@/;
+  const phoneExp = /^\d+(-\d*)*$/;
```
Comment (mmumby) [inline frontend/src/components/search/ContactResultsTable.js]: We have email and phone RegEx's in `Validate.js`
Can we import those and use them in this component? ( I know we spoke in the past about making our email regex the barebones version, so the validate.js can be updated to use the above regEx for email)
- There are also existing checks in that file for phone and email.

File: frontend/src/components/search/MineResultsTable.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        showHeader={false}
+        pagination={false}
+        columns={columns}
+        dataSource={props.searchResults}
+      />
+      {props.showAdvancedLookup && (
+        <Link
+          style={{ float: "right", fontSize: "1.25rem" }}
+          className="padding-large--left"
+          to={router.MINE_HOME_PAGE.dynamicRoute({ search: props.query })}
```
Comment (mmumby) [inline frontend/src/components/search/MineResultsTable.js]: the same question regarding `page` and `per_page` with this route, does this work as expected?

File: frontend/src/components/search/PermitResultsTable.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    {
+      title: "Permit Guid",
+      dataIndex: "permit_guid",
+      key: "permit_guid",
+      render: (text, record) => [
+        <Row>
+          <Col span={24}>
+            <Link
+              to={router.MINE_SUMMARY.dynamicRoute(record.mine_guid, "permit")}
+              style={{ fontSize: "1.25rem" }}
```
Comment (mmumby) [inline frontend/src/components/search/PermitResultsTable.js]: Are these links supposed to be larger than other links inside tables? 
Links should all be the same size, I fear we are going to run into more inconsistencies across components.

File: frontend/src/components/search/SearchBar.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        const newSearchTermHistory = prevState.searchTermHistory.slice(
+          Math.max(prevState.searchTermHistory.length - 2, 0)
+        );
+        newSearchTermHistory.push(searchTerm);
+        return {
+          isSelected: false,
+          searchTermHistory: newSearchTermHistory,
+        };
+      });
+      this.clearSearchBar(false);
```
Comment (mmumby) [inline frontend/src/components/search/SearchBar.js]: `this.clearSearchBar()` can be updated to accept a param, if the param isn't passed in it is implied to be false. 

```
clearSearchBar = (isSelected) => {
    this.props.clearSearchBarResults();
    this.setState({
      isSelected,
      searchTerm: "",
    });
  };
```

File: frontend/src/components/search/SearchResults.js [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+                {`${
+                  type_filter ? this.props.searchOptionsHash[type_filter] : "Search results"
+                } for ${results}`}
+              </h1>
+              <div>
+                {type_filter ? (
+                  <LinkButton
+                    key="all"
+                    onClick={() => {
+                      this.props.history.push(`/search?q=${this.state.params.q}`);
```
Comment (mmumby) [inline frontend/src/components/search/SearchResults.js]: We don't use `this.props.history.push()` anywhere else in the app, 
for consistency, the route should have a dynamic route that we pass the param into

```
export const SEARCH_RESULTS = {
  route: "/search",
dynamicRoute: (q, t) => t ? `/search?q={q}&t={t}` : `/search?q={q}`
  component: SearchResults,
};
```

File: frontend/src/actionCreators/searchActionCreator.js (hunk 7 above)
Comment (nayr974) [reply frontend/src/actionCreators/searchActionCreator.js]: I didn't realize that's how it worked. I'll change that 👍

File: frontend/src/components/search/PermitResultsTable.js (hunk 11 above)
Comment (nayr974) [reply frontend/src/components/search/PermitResultsTable.js]: I thought the wire-frames depicted them larger then the other text, although double checking now they don't. Comparing, I do think it's easier to scan the results with the names being larger but I'll change it for consistency.

File: frontend/src/components/search/ContactResultsTable.js (hunk 9 above)
Comment (nayr974) [reply frontend/src/components/search/ContactResultsTable.js]: Sure, that makes sense!

File: frontend/src/components/search/ContactResultsTable.js (hunk 8 above)
Comment (nayr974) [reply frontend/src/components/search/ContactResultsTable.js]: It does work, although I notice the page count looks wrong. The contact and mine page should probably be responsible for setting its default parameters if hit. I'll see if that's quick to do.

File: frontend/src/components/search/SearchResults.js (hunk 13 above)
Comment (nayr974) [reply frontend/src/components/search/SearchResults.js]: this.props.history.push is used in a few places, although in conjunction with a dynamic route. I'll update it to be in line with the others.

File: frontend/src/actionCreators/searchActionCreator.js (hunk 7 above)
Comment (nayr974) [reply frontend/src/actionCreators/searchActionCreator.js]: Changed!

File: frontend/src/components/search/ContactResultsTable.js (hunk 8 above)
Comment (nayr974) [reply frontend/src/components/search/ContactResultsTable.js]: Added page and per_page params!

File: frontend/src/components/search/ContactResultsTable.js (hunk 9 above)
Comment (nayr974) [reply frontend/src/components/search/ContactResultsTable.js]: Changed!

File: frontend/src/components/search/MineResultsTable.js (hunk 10 above)
Comment (nayr974) [reply frontend/src/components/search/MineResultsTable.js]: Added page and per_page params :)

File: frontend/src/components/search/PermitResultsTable.js (hunk 11 above)
Comment (nayr974) [reply frontend/src/components/search/PermitResultsTable.js]: Changed

File: frontend/src/components/search/SearchResults.js (hunk 13 above)
Comment (nayr974) [reply frontend/src/components/search/SearchResults.js]: Updated 👍

File: frontend/src/components/search/MineResultsTable.js (hunk 10 above)
Comment (mmumby) [reply frontend/src/components/search/MineResultsTable.js]: Thanks, Ryan!

**Pull request merged**